### PR TITLE
Enable bytes serialization

### PR DESCRIPTION
**Pull request by e-kayrakli**
This PR adds serialization/deserialization support to `bytes`.

Resolves https://github.com/Cray/chapel-private/issues/2613.

Specifically:

- Moves some serialization helpers from `String.chpl` to
  `BytesStringCommon.chpl`

- Adds `chpl__serialize`/`chpl__deserialize` to `bytes`.

- Adds new tests.

Test:
- [x] types/bytes
- [x] full

### CODE REVIEW

File: modules/internal/BytesStringCommon.chpl [hunk 1]
```diff
@@ -25,6 +25,27 @@ module BytesStringCommon {
   private use ByteBufferHelpers;
   private use String.NVStringFactory;
 
+  extern const CHPL_SHORT_STRING_SIZE : c_int;
+
+  extern record chpl__inPlaceBuffer {};
+
+  // Signal to the Chapel compiler that the actual argument may be modified.
+  pragma "fn synchronization free"
```
Comment (aconsroe-hpe) [inline modules/internal/BytesStringCommon.chpl:33]: Did you mean to drop the "no doc" here? It also got dropped from the `extern`s and `__serializeHelper`

File: modules/internal/Bytes.chpl [hunk 2]
```diff
@@ -306,6 +306,34 @@ module Bytes {
       }
     }
 
+    proc chpl__serialize() {
```
Comment (aconsroe-hpe) [inline modules/internal/Bytes.chpl:309]: Should these have "no doc"? I didn't see them in `String.chpl` nor do I see them in the generated docs but when I attempted this for regex Elliot suggested adding a no doc.

Comment (e-kayrakli) [reply modules/internal/Bytes.chpl:309]: They don't need to. This is a primary method declared inside a record that has `no doc`. Nothing within the record definition will have documentation generated.

File: modules/internal/BytesStringCommon.chpl (hunk 1 above)
Comment (e-kayrakli) [reply modules/internal/BytesStringCommon.chpl:33]: Yep, and I should drop the one from the next one, as well.

The reason for that is this is a module for which we don't generate any documentation. I see that there are some `no doc`s in this module, but I think they just remained there when I moved the functions from the String module.

(4-5 releases ago, we didn't have the `bytes` type and just `string`. To create the `bytes` type, I created this common module that provides common implementations for the two types. Initially everything was just moved directly from the string module)

File: test/types/bytes/serialization.chpl [hunk 3]
```diff
@@ -0,0 +1,12 @@
+config const myString = "test";
```
Comment (aconsroe-hpe) [inline test/types/bytes/serialization.chpl:1]: Consider adding a test case with long enough string/bytes to actually trigger a GET

**Review by aconsroe-hpe — APPROVED** [review]
With an additional test case to handle longer strings/bytes, this looks good! I'm excited to be able to implement `chpl__serialize` in regex now. Thanks!

**Comment by e-kayrakli** [comment]
@aconsroe-hpe -- Thanks for the review! I merged it already, but can you take a look at the [last commit](https://github.com/chapel-lang/chapel/pull/18666/commits/502c5093f3d23782f51bc18a0178af5eeee45752) to see if the test does more or less what you were thinking of?

**Comment by bradcray** [comment]
Thanks Engin!

**Comment by aconsroe-hpe** [comment]
Yes, it is great! I didn't know you could config a type, I'm totally going to use that in one of my tests!

**Pull request merged**